components/processor/gcomapiprocessor/internal/gcom/cache/instance_cache.go

```diff
@@ -0,0 +1,290 @@
+		err := i.completeCacheRefresh(instanceType)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	go i.run()
+
+	return i, nil
+}
+
+func (i *instanceCache) setLastCacheUpdate(instance client.Instance) {
+	if instance.UpdatedAt != nil && instance.UpdatedAt.After(i.lastCacheUpdate) {
+		i.lastCacheUpdate = *instance.UpdatedAt
+	}
+
+	if instance.CreatedAt.After(i.lastCacheUpdate) {
+		i.lastCacheUpdate = instance.CreatedAt
+	}
+}
+
+func (i *instanceCache) completeCacheRefresh(instanceType client.InstanceType) error {
+	level.Info(i.logger).Log("msg", "attempting to build instance cache", "type", instanceType)
+
+	instanceReq := client.InstanceRequestOptions{
+		Type:     instanceType,
+		PageSize: 1000,
+	}
+
+	if instanceType == client.Grafana {
+		instanceReq.Cluster = i.cfg.GrafanaClusterFilter
+	}
+
+	instances, err := i.gclient.ListInstancesWithPagination(context.TODO(), instanceReq)
+	if err != nil {
+		return err
+	}
+
+	instanceCache := map[int]client.Instance{}
+	idByOrgIDAndName := map[string]int{}
+
+	for _, instance := range instances {
+		i.setLastCacheUpdate(instance)
+		instanceCache[instance.ID] = instance
+		idByOrgIDAndName[strconv.Itoa(instance.OrgID)+instance.Name] = instance.ID
+		level.Debug(i.logger).Log(
+			"msg", "adding instance to cache",
+			"instanceID", instance.ID,
+			"orgID", instance.OrgID,
+			"name", instance.Name)
+	}
+
+	i.cacheMutex.Lock()
+	defer i.cacheMutex.Unlock()
+
+	i.instances[instanceType] = instanceCache
+	i.idByOrgIDAndName[instanceType] = idByOrgIDAndName
+	return nil
+}
+
+func (i *instanceCache) incrementalCacheRefresh(instanceType client.InstanceType) error {
+	level.Info(i.logger).Log("msg", "attempting to build instance cache", "type", instanceType)
+
+	instanceReq := client.InstanceRequestOptions{
+		Type:                  instanceType,
+		PageSize:              1000,
+		IncludeDeleted:        true,
+		UpdatedOrCreatedAtMin: i.lastCacheUpdate,
+	}
+
+	if instanceType == client.Grafana {
+		instanceReq.Cluster = i.cfg.GrafanaClusterFilter
+	}
+
+	instanceList, err := i.gclient.ListInstancesWithPagination(context.TODO(), instanceReq)
+	if err != nil {
+		return err
+	}
+
+	i.cacheMutex.Lock()
+	defer i.cacheMutex.Unlock()
+	for _, instance := range instanceList {
+		i.setLastCacheUpdate(instance)
+
+		_, found := i.instances[instanceType][instance.ID]
+
+		if instance.Status == client.InstanceStatusDeleted {
+			if found {
+				level.Debug(i.logger).Log("msg", "removing deleted instance from cache", "instanceID", instance.ID)
+				delete(i.instances[instanceType], instance.ID)
+			}
+
+			continue
+		}
+
+		level.Debug(i.logger).Log("msg", "adding/updating instance to cache", "instanceID", instance.ID)
+		i.instances[instanceType][instance.ID] = instance
+		i.idByOrgIDAndName[instanceType][strconv.Itoa(instance.OrgID)+instance.Name] = instance.ID
+	}
+
+	return nil
+}
+
+func (i *instanceCache) run() {
+	completeRefreshTicker := time.NewTicker(i.cfg.CompleteCacheRefreshDuration)
```

Comment:
Ideally, this would receive a context, which would be cancelled when the component is shutting down. In that case, the for range channel would be replaced with selects, so that we can cancel the cache refresh operations when we are shutting down.